versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java

```diff
@@ -403,12 +432,55 @@ protected Hash mergeTransplantCommon(
       Consumer<Hash> newKeyLists,
       List<CommitLogEntry> commitsToMergeChronological,
       List<CommitLogEntry> toEntriesReverseChronological,
-      MetadataRewriteParams params)
+      MetadataRewriteParams params,
+      ImmutableMergeResult.Builder<CommitLogEntry> mergeResult)
       throws ReferenceConflictException, ReferenceNotFoundException {
 
     // Collect modified keys.
     Collections.reverse(toEntriesReverseChronological);
-    Set<Key> keysTouchedOnTarget = collectModifiedKeys(toEntriesReverseChronological);
+
+    Map<Key, ImmutableKeyDetails.Builder> keyDetailsMap = new HashMap<>();
+
+    Function<Key, ImmutableKeyDetails.Builder> keyDetails =
+        key ->
+            keyDetailsMap.computeIfAbsent(
+                key,
+                x ->
+                    KeyDetails.builder()
+                        .mergeType(
+                            params
+                                .getMergeTypes()
+                                .getOrDefault(key, params.getDefaultMergeType())));
+
+    BiConsumer<Stream<CommitLogEntry>, BiConsumer<ImmutableKeyDetails.Builder, Iterable<Hash>>>
+        keysFromCommitsToKeyDetails =
+            (commits, receiver) -> {
+              Map<Key, Set<Hash>> keyHashesMap = new HashMap<>();
+              Function<Key, Set<Hash>> keyHashes =
+                  key -> keyHashesMap.computeIfAbsent(key, x -> new LinkedHashSet<>());
+              commits.forEach(
+                  commit -> {
+                    commit
+                        .getDeletes()
+                        .forEach(delete -> keyHashes.apply(delete).add(commit.getHash()));
+                    commit
+                        .getPuts()
+                        .forEach(put -> keyHashes.apply(put.getKey()).add(commit.getHash()));
+                  });
+              keyHashesMap.forEach((key, hashes) -> receiver.accept(keyDetails.apply(key), hashes));
+            };
+
+    Set<Key> keysTouchedOnTarget = new HashSet<>();
+    keysFromCommitsToKeyDetails.accept(
+        commitsToMergeChronological.stream(), ImmutableKeyDetails.Builder::addAllSourceCommits);
+    keysFromCommitsToKeyDetails.accept(
+        toEntriesReverseChronological.stream()
+            .peek(
+                e -> {
+                  e.getPuts().stream().map(KeyWithBytes::getKey).forEach(keysTouchedOnTarget::add);
+                  e.getDeletes().forEach(keysTouchedOnTarget::remove);
```

Comment:
We do expose the commit log entries including the operations.
The idea is that you have the (conflicting) commit-IDs per key and then look up the commit log entries, which hold the actual operations.

Reply:
One caveat w/ the "target commits" in `MergeResult`: those "target commits" are only the commits since expected-hash up to the branch's head. For a "transplant" we can't do more than that. For a "merge" we could, but the number of commits on the target branch could be quite a lot, maybe too many - think of having an experiment branch running for several weeks - there could be a huge amount of commits on the target branch.

Reply:
I mean if we have a `Delete A` in `main` and a `Put A` in `test` , then merge `test`, my reading that our code will not treat this as a conflict, but I think it should. WDYT?

Reply:
(I know this PR does not change the conflict login, I guess noticed this while reviewing)

Reply:
Right, that's not a conflict. I'm not sure whether re-adding a table from a merged branch should be a conflict.